shared_model/utils/lazy_initializer.hpp

```diff
@@ -35,18 +35,17 @@ namespace shared_model {
       using GeneratorType = std::function<Target()>;
 
      public:
-      explicit LazyInitializer(const GeneratorType &generator)
-          : generator_(generator) {}
-
-      LazyInitializer(const LazyInitializer &) = default;
+      template <typename T>
+      explicit LazyInitializer(T &&generator)
+          : generator_(std::forward<T>(generator)) {}
 
       using PointerType = typename std::add_pointer_t<Target>;
 
       const Target &operator*() const { return *ptr(); }
 
       const PointerType ptr() const {
         if (target_value_ == boost::none) {
-          target_value_ = boost::make_optional<Target>(generator_());
+          target_value_.emplace(generator_());
```

Comment:
Not an issue, what `emplace` means here?

Reply:
Same as [in stl](http://en.cppreference.com/w/cpp/container/vector/emplace), I guess

Reply:
This is due to using `operator=` with `boost::optional<Target>` requires copy constructor to be present, which is not intended here.